src/session/session_api.c

```diff
@@ -1596,31 +1618,53 @@ __wt_session_range_truncate(
         WT_ERR(__session_open_cursor((WT_SESSION *)session, stop->uri, NULL, NULL, &start));
         local_start = true;
         WT_ERR(start->next(start));
+        /* Record new start cursor. */
+        trunc_info->start = start;
     }
 
     /*
      * If the start/stop keys cross, we're done, the range must be empty.
      */
     if (stop != NULL) {
         WT_ERR(start->compare(start, stop, &cmp));
-        if (cmp > 0)
+        if (cmp > 0) {
+            log_trunc = true;
             goto done;
+        }
     }
 
-    /*
-     * Now that the truncate is setup and ready regardless of how the API was called, populate our
-     * truncate information cookie.
-     */
-    trunc_info->session = session;
-    trunc_info->uri = uri != NULL ? uri : start->internal_uri;
-    trunc_info->start = start;
-    trunc_info->stop = stop;
-    trunc_info->orig_start_key = orig_start_key;
-    trunc_info->orig_stop_key = orig_stop_key;
-
     WT_ERR(__wt_schema_range_truncate(trunc_info));
 
 done:
+    /*
+     * In the cases where truncate doesn't have work to do, we still need to generate a log record
+     * for the operation. That way we can be consistent with other competing inserts or truncates on
+     * other tables in this transaction.
+     */
+    if (log_trunc) {
+        /*
+         * If we have cursors and know there is no work to do, there may not be a dhandle in the
+         * session. Grab it from the start or stop cursor as needed.
+         */
+        dhandle = session->dhandle;
+        if (dhandle == NULL && start != NULL) {
+            cbt = (WT_CURSOR_BTREE *)start;
+            dhandle = cbt->dhandle;
+        } else if (dhandle == NULL && stop != NULL) {
+            cbt = (WT_CURSOR_BTREE *)stop;
+            dhandle = cbt->dhandle;
+        }
+        /* We have to have a dhandle from somewhere. */
+        WT_ASSERT(session, dhandle != NULL);
```

Comment:
I'm not sure I care, but won't this mean that no-op range truncate of complex tables won't log the truncate operation? Will that lead complex tables to being susceptible to the bug?

Reply:
I believe that is correct. How about I open a new ticket for complex table logging and not tie it into this ticket?

Reply:
I created WT-11059 for the complex table work.